[LBSE] Incorrect clipping for non-opaque, outermost <svg> elements

nikolaszimmermann opened the pull request:
#### 7df9f6186a3b0f0341ef887829fc8b1f33dbdda2
<pre>
[LBSE] Incorrect clipping for non-opaque, outermost &lt;svg&gt; elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=262843">https://bugs.webkit.org/show_bug.cgi?id=262843</a>

Reviewed by Rob Buis.

RenderLayer::beginTransparencyLayers() pixel snaps the clip rect before
clipping, which shouldn&apos;t be done for SVG elements. Other places already
use &apos;snapRectToDevicePixelsIfNeeded()&apos;, this one was still using
&apos;snapRectToDevicePixels()&apos; unconditionally.

Removes the need for a bunch of LBSE specific pixel test results (where the
tests pass, except for opacity clipping related glitches that are now fixed).

Covered by existing tests.

* LayoutTests/platform/mac-ventura-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/render-groups-01-b-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/as-background-image/svg-as-background-6-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/batik/text/textAnchor-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/container-opacity-clip-viewBox-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/js-update-bounce-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/scrolling-embedded-svg-file-image-repaint-problem-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-pixel/svg/as-background-image/svg-as-background-6-expected.png:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::beginTransparencyLayers): Only pixel snap the clip rect, if needed.
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::clippedOverflowRect const): Deleted. No longer needed, use the RenderReplaced variant.
* Source/WebCore/rendering/svg/RenderSVGRoot.h:

Canonical link: <a href="https://commits.webkit.org/269360@main">https://commits.webkit.org/269360@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6dbb10165163116e71d75f4572a20cd48146dd3f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21444 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21749 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22497 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23306 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19873 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21685 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/25051 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21998 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21068 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21669 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21309 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18587 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24157 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18488 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19437 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25750 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19573 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19678 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23599 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20124 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17138 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19442 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5319 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23682 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20031 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/rendering/svg/RenderSVGRoot.h:
```diff
@@ -108,7 +108,6 @@ class RenderSVGRoot final : public RenderReplaced {
     bool nodeAtPoint(const HitTestRequest&, HitTestResult&, const HitTestLocation& locationInContainer, const LayoutPoint& accumulatedOffset, HitTestAction) final;
 
     LayoutRect overflowClipRect(const LayoutPoint& location, RenderFragmentContainer* = nullptr, OverlayScrollbarSizeRelevancy = IgnoreOverlayScrollbarSize, PaintPhase = PaintPhase::BlockBackground) const final;
-    LayoutRect clippedOverflowRect(const RenderLayerModelObject* repaintContainer, VisibleRectContext) const final;
```
Why is it safe to remove this?

nikolaszimmermann replied:
There is no deviation anymore with RenderReplaced -- no need for RenderSVGRoot to keep around this method.

rwlbuis replied:
Nice!

rwlbuis reviewed (approved):
LGTM.

nikolaszimmermann committed:
[LBSE] Incorrect clipping for non-opaque, outermost <svg> elements
https://bugs.webkit.org/show_bug.cgi?id=262843

Reviewed by Rob Buis.

RenderLayer::beginTransparencyLayers() pixel snaps the clip rect before
clipping, which shouldn't be done for SVG elements. Other places already
use 'snapRectToDevicePixelsIfNeeded()', this one was still using
'snapRectToDevicePixels()' unconditionally.

Removes the need for a bunch of LBSE specific pixel test results (where the
tests pass, except for opacity clipping related glitches that are now fixed).

Covered by existing tests.

* LayoutTests/platform/mac-ventura-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/render-groups-01-b-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/as-background-image/svg-as-background-6-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/batik/text/textAnchor-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/container-opacity-clip-viewBox-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/js-update-bounce-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/scrolling-embedded-svg-file-image-repaint-problem-expected.png: Removed.
* LayoutTests/platform/mac-ventura-wk2-pixel/svg/as-background-image/svg-as-background-6-expected.png:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::beginTransparencyLayers): Only pixel snap the clip rect, if needed.
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::clippedOverflowRect const): Deleted. No longer needed, use the RenderReplaced variant.
* Source/WebCore/rendering/svg/RenderSVGRoot.h:

Canonical link: https://commits.webkit.org/269360@main
